femwell/maxwell/waveguide.py

```diff
@@ -148,6 +149,27 @@ def calculate_pertubated_neff(self, delta_epsilon):
             * 0.5
         )
 
+    def calculate_intensity(self, field: NDArray) -> Tuple[NDArray, Basis]:
+        """Calculates the intensity of a field as the sum of the absolute values squared of its components.
+
+        First, the field is interpolated on the quadrature points with a discontinuous
+        element, then the intensity is calculated directly on the quadrature points.
+
+        Args:
+            field (NDArray): Field whose intensity is to be calculated.
+
+        Returns:
+            NDArray, Basis: Interpolated intensity and plot-ready basis
+        """
+        (et, et_basis), (ez, ez_basis) = self.basis.split(field)
+        plot_basis = et_basis.with_element(ElementVector(ElementDG(ElementTriP1())))
+        et_xy = plot_basis.project(et_basis.interpolate(et), dtype=np.complex128)
+        (et_x, et_x_basis), (et_y, et_y_basis) = plot_basis.split(et_xy)
+        ez2 = et_x_basis.project(ez_basis.interpolate(ez), dtype=np.complex128)
+        intensity = np.abs(et_x) ** 2 + np.abs(et_y) ** 2 + np.abs(ez2) ** 2
```

Comment:
I think like this the calculation will give the right result, but it's totally not executed as it should be, sorry :(

How it should be done:
- interpolate the field on the quadrature points
- square there each field individually and sum the fields up
- project back on the ElementDG(ElementTriP1())-Basis

Such calculations should normally only be done at the quadrature points (I think linear operations are okay to do on the DOFs directly)
For a linear Basis, it works like this, as the DOFs represent the values at the nodes, but I think it's better to stay consistently (especially if we one day want a higher accurrancy interpolation, for a quadratic Basis it won't work like this)

Reply:
I see, thanks for the feedback! Will be corrected soon :smiley: